feat: article on exporting PDFs (#1706)
stoplightio/stoplight.io: netlify/blog-posts/pdf-api-documentation.md

tags:
  - blog
  - blog-featured
  - blog-general
relatedTags:
  - blog
  - blog-general
publishedDate: 2021-01-04
author: Phil Sturgeon
title: Export API Documentation as a PDF
subtitle: >-
  Using OpenAPI to create PDF based reference documentation for clients and other third-parties.
image: /images/blog-images/documentation-cover.png
includeToc: true
actionBar:
  ctas:
    - color: blue2
      href: 'https://stoplight.io/api-documentation'
      title: Get Started
      large: true
  text: Create comprehensive documentation with Stoplight.
meta:
  description: >-
    Using OpenAPI to create PDF based reference documentation for clients and other third-parties
  favicon: /images/mark_light_bg.png
  robots: 'index, follow'
  title: >-
    Export API Documentation as a PDF
  image: /images/blog-images/openapi-to-pdf.png
  url: https://stoplight.io/blog/pdf-api-documentation/
  twitter:
    description: >-
      Using OpenAPI to create PDF based reference documentation for clients and other third-parties
    title: >-
      Export API Documentation as a PDF
    image: /images/blog-images/openapi-to-pdf.png
    url: https://stoplight.io/blog/pdf-api-documentation/
---

> You've built a brilliant API, well done! But hold up, people are trying to use it, and there's no public facing documentation. Your API and the docs are private, and require a special VPN, SAML, and/or some sort of OAuth proxy that requires not just a GitHub account, but that GitHub user needs to be part of your organization.

Reviewer: ```suggestion
You've built a brilliant API, well done! But hold up, people are trying to use it, and there's no public-facing documentation. Your API and the docs are private, and require a special VPN, SAML, and/or some sort of OAuth proxy that requires not just a GitHub account, but that GitHub user needs to be part of your organization.
```

> Whatever the approach, there are often users outside of this super-secure realm that need access to the documentation. Maybe it's a customer who is paying a bunch to get access to some of your data. Maybe its a contractor who is going to be doing some work on the API and doesn't want to [guess the contract](https://apisyouwonthate.com/blog/commit-to-api-contracts). Maybe it's a different department who are in a different GitHub organization, e.g.: an acquired company wants to integrate.

Reviewer: ```suggestion
Whatever the approach, there are often users outside of this super-secure realm that needs access to the documentation. Maybe it's a customer who is paying a bunch to get access to some of your data. Maybe it's a contractor who is going to be doing some work on the API and doesn't want to [guess the contract](https://apisyouwonthate.com/blog/commit-to-api-contracts). Maybe it's a different department that is in a different GitHub organization, e.g.: an acquired company wants to integrate.
```

So what do you do? A common approach is to export the API docs as a PDF, then email it over. Sure. That'll work, but _how_?!

## Step 1: Create an OpenAPI Description

For a lot of you reading this, you'll already have an OpenAPI description. Someone on your team probably [created it before anyone started programming the API](https://stoplight.io/api-design-guide/basics/#why-api-design-first-matters), so that everyone could be sure the right thing was being built in the first place. Other teams might have [generated OpenAPI using annotations](https://stoplight.io/blog/generate-documentation-from-code/) added to the code.

If you've not got any OpenAPI yet head over to [Stoplight Studio](/studio) or look for another [Visual OpenAPI Editor](https://openapi.tools/#gui-editors) to put together some OpenAPI for your API. Each PDF conversion approach will have an example you can try.

> ## Step 2: Convert OpenAPI to PDF

Reviewer: ```suggestion
## Step 2: Convert to PDF
```

### Print from Browser 🤞

Just to get this one out of the way, some folks recommending printing their existing API documentation. They load it up in their browser, and hit "Print > Save as PDF". This is not likely to give great results.

The popular API documentation tools use the "three-column" layout, popularized by [Stripe API Reference](https://stripe.com/docs/api) with a navigation bar on the left and other stuff on the right, which looks daft squashed onto an A4 page.

Then there's the interactive elements in most good API documentation, like interactive HTTP consoles, parameter lists with expandable nested content, and generally a bunch of hyperlinks that don't work when you print them. 

![Stripe often considered the best API docs around, look rather awkward and unusable when printed due to collapsed nested objects not being expandable.](/images/blog/pdf-api-documentation/print-comparison.jpg)

If your API docs are using a really clever `print.css` that hides everything not useful in a print context, and expanded everything that should be expanded, it might just work out fine. This is rare, but worth trying. If it doesn't work, try another one of these approaches.

> ### swagger2pdf

Reviewer: ```suggestion
### Swagger2pdf
```

Looks weird with other headings

<html>
  <head>
    <script src="https://unpkg.com/rapipdf/dist/rapipdf-min.js"></script>
  </head>
  <body>
    <rapi-pdf
      style = "width:700px; height:40px; font-size:18px;"
      spec-url =
      "https://stoplightio.github.io/sample-specs/reference/instagram/instagram.yaml"
    > </rapi-pdf>
  </body>
</html>
```

Then open that file in the browser (e.g.: `$ open generate.html`) and click the Generate button to trigger a download of the PDF.

RapiPDF creates a more useful output, especially for schemas with a lot of nesting (objects inside objects), and doesn't contain any links or elements that shouldn't be in a PDF.

![Screenshot of a POST request in the PDF produced by swagger2pdf](/images/blog/pdf-api-documentation/rapipdf.png)

As this one is a bit tricky to show off in a single screenshot, take a look at [the resulting PDF here](/images/blog/pdf-api-documentation/rapipdf-instagram-example.pdf).

## Alternative Workflows

Instead of doing any of this, it might be worth taking a step back, and thinking "Why do I need to make a PDF?" Somebody may have asked you to do it, or you've decided to do it because it's something you've always done, but... do you need to?

> The base use-case here is often an API that may or many not be private, but definitely has private or "internal" documentation, and somebody who is not in that "in group" needs access to it. There are a lot of other ways to do this. 

Reviewer: ```suggestion
The base use-case here is often an API that may or may not be private, but definitely has private or "internal" documentation, and somebody who is not in that "in group" needs access to it. There are a lot of other ways to do this. 
```

For starters, you could generate HTML with any of the many free and open-source tools that create [Beautiful API Reference Documentation](https://apisyouwonthate.com/blog/turning-contracts-into-beautiful-documentation), and host that on a [password-protected Amazon S3 bucket](https://kynatro.com/blog/2018/01/03/a-step-by-step-guide-to-creating-a-password-protected-s3-bucket/) then email the password to and a URL instead of emailing a PDF. If the S3 bucket is updated with new HTML by your continuous integration server then you don't have to worry about emailing over updated PDFs.

> If you don't want to set up _any_ of this, Stoplight handles the workflow for you out of thr box. There's no need to email PDFs around, or roll your own password-protected hosted documentation. We already have private hosted documentation based on SAML, and version-control-based SSO from GitHub, BitBucket, etc., then you can whitelist email address domains too. 

Reviewer: ```suggestion
If you don't want to set up _any_ of this, Stoplight handles the workflow for you out of the box. There's no need to email PDFs around, or roll your own password-protected hosted documentation. We already have private hosted documentation based on SAML, and version-control-based SSO from GitHub, BitBucket, etc., then you can whitelist email address domains too. 
```

> Maybe you work at `protectedcompany.com` and you want anyone at `somecustomer.com` to get access to certain projects? Whitelist `somecustomer.com` with Guest priviledges, get them to sign up, and add them to the specific projects they should be allowed to see. Maybe you have a contractor who should be treated like an employee and have access to seeing most APIs? Make them a Viewer and now they can see any internal projects.

Reviewer: ```suggestion
Maybe you work at `protectedcompany.com` and you want anyone at `somecustomer.com` to get access to certain projects? Whitelist `somecustomer.com` with Guest privileges, get them to sign-up, and add them to the specific projects they should be allowed to see. Maybe you have a contractor who should be treated like an employee and have access to seeing most APIs? Make them a Viewer and now they can see any internal projects.
```
